commit 8c087b5183adab186a298f2d6ed39aefdcae413c
Author: Chris Ball <cjb@laptop.org>
Date:   Wed Feb 4 09:29:54 2009 -0500

    Btrfs: Handle SGID bit when creating inodes
    Before this patch, new files/dirs would ignore the SGID bit on their
    parent directory and always be owned by the creating user's uid/gid.
    
    Signed-off-by: Chris Ball <cjb@laptop.org>
    Signed-off-by: Chris Mason <chris.mason@oracle.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index 95ea58c..5792816 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -3472,7 +3472,14 @@ static struct inode *btrfs_new_inode(struct btrfs_trans_handle *trans,
 		root->highest_inode = objectid;
 
 	inode->i_uid = current_fsuid();
-	inode->i_gid = current_fsgid();
+
+	if (dir->i_mode & S_ISGID) {
+		inode->i_gid = dir->i_gid;
+		if (S_ISDIR(mode))
+			mode |= S_ISGID;
+	} else
+		inode->i_gid = current_fsgid();
+
 	inode->i_mode = mode;
 	inode->i_ino = objectid;
 	inode_set_bytes(inode, 0);


